### DISCUSSION TITLE
Standardization

### DISCUSSION THREAD

**Discussion by zachhardy** [chi-tech/chi-tech#283] [Ideas]
I'd like to open a dialogue about standardizing some coding rules going forward so that there is more consistent syntax and style within files. I talked about this with @Naktakala a few days back and came up with a list of items we should think about:

- Rules for when to inline functions and whether the the definition should be at the declaration or lower in the header files. 
- Rules for what needs to be documented, where documentation should be, and what particular types of documentation should include.
    - For example, when is it okay to not document a function? Where do descriptions of parameters go? Where does the description of the method go?
-  Rules for how to declare classes.
    -  Should attributes be declared before routines or vice versa? Should there be some ordering from public to protected to private? And so on.
- Rules for when something should be public, protected, or private and naming conventions.
    - For example, do we want to prepend protected/private members with "m_" or use a trailing "_"?
    - Should we have as much stuff public as we do?
- Should we use "and", "or", and "not" or "&&", "||", "!"?
- When should we use "int" vs. "unsigned int" vs. "size_t" in for loops?

This is obviously not a task that is feasible to handle right now. But over time as files change and as new files are created, it would be good to gradually standardize them. When I have the time, I will try to write up some guidelines in the documentation that we can iterate back and forth on.

**Comment by ragusa** [comment]
I've always looked at what other projects do:

https://dealii.org/developer/developers/writing-documentation.html

https://dealii.org/developer/doxygen/deal.II/CodingConventions.html

https://mooseframework.inl.gov/sqa/framework_scs.html

https://docs.openmc.org/en/stable/devguide/index.html

I'm up for any improvements!
